src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
+        else 
+        {
+          byte evindex = msg->data[5];
+          byte evval = msg->data[6];
+          // search for this NN, EN as we may just be adding an EV to an existing learned event
+          // DEBUG_SERIAL << F("> searching for existing event to update") << endl;
+          byte index = module_config->findExistingEvent(nn, en);
+
+          // not found - it's a new event
+          if (index >= module_config->EE_MAX_EVENTS) 
+          {
+            // DEBUG_SERIAL << F("> existing event not found - creating a new one if space available") << endl;
+            index = module_config->findEventSpace();
+          }
+
+          // if existing or new event space found, write the event data
+          if (index < module_config->EE_MAX_EVENTS) 
+          {
+            // write the event to EEPROM at this location -- EVs are indexed from 1 but storage offsets start at zero !!
+            // DEBUG_SERIAL << F("> writing EV = ") << evindex << F(", at index = ") << index << F(", offset = ") << (module_config->EE_EVENTS_START + (index * module_config->EE_BYTES_PER_EVENT)) << endl;
+
+            // don't repeat this for subsequent EVs
+            if (evindex < 2) 
+            {
+              module_config->writeEvent(index, &msg->data[1]);
+            }
+
+            module_config->writeEventEV(index, evindex, evval);
+
+            // recreate event hash table entry
+            // DEBUG_SERIAL << F("> updating hash table entry for idx = ") << index << endl;
+            module_config->updateEvHashEntry(index);
+
+            // respond with WRACK
+            controller->sendWRACK();  // Deprecated in favour of GRSP_OK
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), GRSP_OK);
+          } 
+          else 
+          {
+            // DEBUG_SERIAL << F("> no free event storage, index = ") << index << endl;
+            // respond with CMDERR & GRSP
+            controller->sendCMDERR(CMDERR_TOO_MANY_EVENTS);
+            controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_TOO_MANY_EVENTS);
+          }
+        }
       }
 
-      // if existing or new event space found, write the event data
-      if (index < module_config->EE_MAX_EVENTS)
-      {
-        // write the event to EEPROM at this location -- EVs are indexed from 1 but storage offsets start at zero !!
-        // DEBUG_SERIAL << F("> writing EV = ") << evindex << F(", at index = ") << index << F(", offset = ") << (module_config->EE_EVENTS_START + (index * module_config->EE_BYTES_PER_EVENT)) << endl;
+      return PROCESSED;
 
-        // don't repeat this for subsequent EVs
-        if (evindex < 2)
+    case OPC_EVLRNI:
+      // F5 - learn an event using event index
+
+      DEBUG_SERIAL << endl
+                   << F("> EVLRNI for source nn = ") << nn << endl;
+
+      // we must be in learn mode
+      if (bLearn) 
+      {
+        if (msg->len < 8) 
+        {
+          controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_INV_CMD);
```

Comment:
OP-code spec says nothing about GRSP for any of the error cases. I guess because it is deprecated in VLCB.

Reply:
I think that the spec should be updated to put in these error messages, particularly as I don't think that this op code should be deprecated.  I have found it useful for teaching short events to producers.  I've added it to the agenda for the next meeting.

Reply:
OK.